[TableGen] Add `!instances` operator to get defined records

wangpc-pp opened the pull request:
The format is: `!instances<T>([regex])`.
    
This operator produces a list of records whose type is `T`. If
`regex` is provided, only records whose name matches the regular
expression `regex` will be included. The format of `regex` is ERE
(Extended POSIX Regular Expressions).

wangpc-pp commented:
Add some RISC-V reviewers here because the intention is to simplify the generation of lists of RVV pseudos. For example:
```
list<RISCVVPseudo> VADD = !select<RISCVVPseudo>(".*VADD.*");
list<RISCVVPseudo> M8 = !select<RISCVVPseudo>(".*_M8");
```

jayfoad commented:
My gut feeling is that this operator does too much, and the name "select" is too generic. Can you split the functionality into smaller pieces:
- generate a list of names of records of type T
- filter a list based on a regex (using `!filter`?)
- cast the names back to records (using `!foreach` and `!cast`?)

jayfoad commented:
> 1. For the first problem, we can add a operator `!defined<T>()` which is a wrapper of `getAllDerivedDefinitionsIfDefined`.

Sounds reasonable to me.

> 2. For the second problem, we can add an operator `!match(str, regex)` to test if `str` matches `regex`.

Sounds reasonable, although there might be arguments about exactly what regex syntax it should support.

> 3. To simplify `!foreach(!filter(!match(), !defined()), !cast)` chain (and optimize the performance since it takes too many intermediate steps), the `!defined` can accept an optional regex parameter and the format becomes `!defined<T>(regex?)`.

This seems like premature optimization. Is there really a significant performance problem?

wangpc-pp commented:
> > 1. For the first problem, we can add a operator `!defined<T>()` which is a wrapper of `getAllDerivedDefinitionsIfDefined`.
> 
> Sounds reasonable to me.
> 
> > 2. For the second problem, we can add an operator `!match(str, regex)` to test if `str` matches `regex`.
> 
> Sounds reasonable, although there might be arguments about exactly what regex syntax it should support.

I just use `llvm/include/llvm/Support/Regex.h` so it is a POSIX one: `This file implements a POSIX regular expression matcher. Both Basic and Extended POSIX regular expressions (ERE) are supported.`

> 
> > 3. To simplify `!foreach(!filter(!match(), !defined()), !cast)` chain (and optimize the performance since it takes too many intermediate steps), the `!defined` can accept an optional regex parameter and the format becomes `!defined<T>(regex?)`.
> 
> This seems like premature optimization. Is there really a significant performance problem?

I strongly believe it is! The TableGen chained operators will be at least 100x slower than the native filter implemented by C++.

topperc commented:
> Add some RISC-V reviewers here because the intention is to simplify the generation of lists of RVV pseudos. For example:
> 
> ```python
> list<RISCVVPseudo> VADD = !select<RISCVVPseudo>(".*VADD.*");
> list<RISCVVPseudo> M8 = !select<RISCVVPseudo>(".*_M8");
> ```

Where do we need lists like this?

arsenm commented:
> This seems like premature optimization. Is there really a significant performance problem?

It's tablegen, so when isn't there one?

wangpc-pp committed:
[TableGen] Add `!records` operator to get defined records

The format is: `!records<T>([regex])`.

This operator produces a list of records whose type is `T`. If
`regex` is provided, only records whose name matches the regular
expression `regex` will be included. The format of `regex` is ERE
(Extended POSIX Regular Expressions).

wangpc-pp commented:
Ping!

jayfoad commented:
I don't like the name `!defined`. Maybe `!instances` would be better? Anyone else have ideas?

jayfoad commented on llvm/docs/TableGen/ProgRef.rst:
```diff
@@ -1722,6 +1722,11 @@ and non-0 as true.
+``!defined<``\ *type*\ ``>([``\ *regex*\ ``])``
+    This operator produces a list of records whose type is *type*. If *regex*
+    is provided, only records whose name matches the regular expression *regex*
+    will be included.
```
You should document what syntax the regex uses.

wangpc-pp commented:
> I don't like the name `!defined`. Maybe `!instances` would be better? Anyone else have ideas?

Maybe `records`? `record` is the terminology used in https://llvm.org/docs/TableGen/ProgRef.html.

jayfoad commented:
> > I don't like the name `!defined`. Maybe `!instances` would be better? Anyone else have ideas?
> 
> Maybe `records`? `record` is the terminology used in https://llvm.org/docs/TableGen/ProgRef.html.

As an aside about naming: I know that record are called "records". But the name of this operator should suggest not just "records"; it should suggest "list of all records derived from a specified class".

wangpc-pp commented:
> > > I don't like the name `!defined`. Maybe `!instances` would be better? Anyone else have ideas?
> > 
> > 
> > Maybe `records`? `record` is the terminology used in https://llvm.org/docs/TableGen/ProgRef.html.
> 
> As an aside about naming: I know that record are called "records". But the name of this operator should suggest not just "records"; it should suggest "list of all records derived from a specified class".

Yeah, what about `recordsof`?

jayfoad commented:
I _think_ I still prefer `!instances`, but I would be happy to hear any other suggestions or opinions!

wangpc-pp committed:
Rename to instances

wangpc-pp commented:
> I _think_ I still prefer `!instances`, but I would be happy to hear any other suggestions or opinions!

Make sense to me, I renamed it to `!instances`!

Paul-C-Anagnostopoulos commented:
> I _think_ I still prefer `!instances`, but I would be happy to hear any other suggestions or opinions!

Yes, !instances.

jurahul commented on llvm/include/llvm/TableGen/Record.h:
```diff
@@ -1192,6 +1193,41 @@ class ExistsOpInit final : public TypedInit, public FoldingSetNode {
+class InstancesOpInit final : public TypedInit, public FoldingSetNode {
+private:
+  const RecTy *Type;
+  const Init *Regex;
```
Can this be a `StringInit` here? That way you avoid an error check and prevent creating an illegal InstancesOpInit instance.

wangpc-pp replied:
No we can't. the `Regex` can be any `Init` with type `StringRecTy` and it may not be resolved when folding.
Similar comments before: https://github.com/llvm/llvm-project/pull/130759#discussion_r1991086241

jurahul commented on llvm/lib/TableGen/TGParser.cpp:
```diff
@@ -1455,6 +1455,49 @@ const Init *TGParser::ParseOperation(Record *CurRec, const RecTy *ItemType) {
+      }
+    } else {
+      // Use wildcard when Regex is not specified.
+      Regex = StringInit::get(Records, ".*");
```
wondering if this case can he handled inside the Fold method instead and we use a null StringInit in the constructor. The only reason is that get `InstancesOpInit::getAsString` in this case will print something different than what was in the input (which maybe ok?).

wangpc-pp replied:
Yeah, my initial implementation was like what you suggested. I dropped that and switched to this way because this can simplify a lot parsing/folding code. I think the output of print looks good. :-)

Artem-B commented on llvm/docs/TableGen/ProgRef.rst:
```diff
@@ -1836,6 +1836,12 @@ and non-0 as true.
   This operator produces 1 if *a* is not the uninitialized value (``?``) and 0
   otherwise.
 
+``!instances<``\ *type*\ ``>([``\ *regex*\ ``])``
+    This operator produces a list of records whose type is *type*. If *regex*
+    is provided, only records whose name matches the regular expression *regex*
+    will be included. The format of *regex* is ERE (Extended POSIX Regular
+    Expressions).
```
At which point in time will the existence of those instances be considered?

See more of my questions in the test file below.

wangpc-pp replied:
Good point! I will add more tests and clarify the rule of visibility.

Artem-B commented on llvm/test/TableGen/records.td:
```diff
@@ -0,0 +1,60 @@
+// CHECK-LABEL: def instances_B {
+// CHECK-NEXT:    list<B> instances = [b0, b1];
+// CHECK-NEXT:  }
+
```
If I add `def a3: A` here, will it be included in `instances_A` above? Or only in the subsequent uses of `!instances<A>()` ?

wangpc-pp replied:
The answer is a3 will not be included.

Artem-B commented on llvm/test/TableGen/records.td:
```diff
@@ -0,0 +1,60 @@
+class B : A;
+def b0 : B;
+def b1 : B;
+
```
If I use `!instances<>` inside of a class, does it take effect at the point where it's parsed, or at the point where the class record is instantiated?

```
class X {
   list<A> instances = !instances<A>();
}

def ax: A;
def x: X; 
```

Is such use allowed at all? If so, will `x.instances` include `ax`?

We do not seem to have any test cases for use of `!instances` inside class/multiclass. We want to check use cases in the body and in the default arguments.

wangpc-pp replied:
I added more tests. The visibility is at the point where the class record is instantiated.

wangpc-pp committed:
Rename records.td to instances.td

wangpc-pp committed:
Add more tests and clarify the visibility

Artem-B commented on llvm/test/TableGen/instances.td:
```diff
@@ -0,0 +1,134 @@
+// CHECK-NEXT:  }
+def test7_default_arg : TestArgument;
+
+// Temporary actual parameter won't be considered.
```
This sounds counter-intuitive. For the instantiation of `TestArgument` both default or explicitly provided argument look exectly the same -- the value record exists in both acases. Capturing it in case of the value coming from the default argument, but not for the actual argument looks inconsistent to me. I think behavior should be the same in both cases.

wangpc-pp replied:
Thanks, good catch! I have the same suspect, it is because of the Init cache mechanism.

wangpc-pp committed:
Rework argument tests

Artem-B reviewed (approved):
LGTM.
